### PR TITLE
bulk indexing

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: batch-index/build.gradle [hunk 1]
```diff
@@ -11,6 +11,8 @@ dependencies {
     implementation group: 'software.amazon.awssdk', name: 'url-connection-client', version: project.ext.awsSdk2Version
     implementation group: 'software.amazon.awssdk', name: 'eventbridge', version:project.ext.awsSdk2Version
 
+    implementation group: 'com.google.guava', name: 'guava', version: '30.1.1-jre'
```
Comment (axthosarouris) [inline batch-index/build.gradle:14]: for splitting a list into Lists

File: search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 2]
```diff
@@ -134,14 +140,29 @@ public void removeDocumentFromIndex(String identifier) throws SearchException {
 
     public AcknowledgedResponse prepareIndexForBatchInsert() throws IOException {
         Settings indexSettings = Settings.builder()
-                                     .put(ELASTIC_SEARCH_INDEX_REFRESH_INTERVAL, FIFTEEN_MINUTES)
-                                     .put(ELASTIC_SEARCH_NUMBER_OF_REPLICAS, 0)
-                                     .build();
+            .put(ELASTIC_SEARCH_INDEX_REFRESH_INTERVAL, FIFTEEN_MINUTES)
```
Comment (axthosarouris) [inline search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:143]: autoformating

File: search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 3]
```diff
@@ -134,14 +140,29 @@ public void removeDocumentFromIndex(String identifier) throws SearchException {
...
+            .put(ELASTIC_SEARCH_NUMBER_OF_REPLICAS, 0)
+            .build();
         return indexExists()
                    ? updateCurrentIndex(indexSettings)
                    : createNewIndexWithOptimizedSettings(indexSettings);
     }
 
+    public List<BulkResponse> batchInsert(List<IndexDocument> indexDocuments) {
```
Comment (axthosarouris) [inline search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:151]: this is the main change. we break the list of index documents to bulks and send bulks for indexing

File: search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 4]
```diff
@@ -151,9 +172,9 @@ protected final RestHighLevelClientWrapper createElasticsearchClientWithIntercep
                                                    credentialsProvider);
 
         RestClientBuilder clientBuilder = RestClient
-                                              .builder(HttpHost.create(ELASTICSEARCH_ENDPOINT_ADDRESS))
-                                              .setHttpClientConfigCallback(
-                                                  hacb -> hacb.addInterceptorLast(interceptor));
+            .builder(HttpHost.create(ELASTICSEARCH_ENDPOINT_ADDRESS))
```
Comment (axthosarouris) [inline search-commons/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:175]: autoformating

File: search-commons/src/main/java/no/unit/nva/search/RestHighLevelClientWrapper.java [hunk 5]
```diff
@@ -67,4 +69,9 @@ public UpdateResponse update(UpdateRequest updateRequest, RequestOptions request
     public IndicesClientWrapper indices() {
         return new IndicesClientWrapper(client.indices());
     }
+
+    @JacocoGenerated
+    public BulkResponse bulk(BulkRequest request, RequestOptions requestOption) throws IOException {
```
Comment (axthosarouris) [inline search-commons/src/main/java/no/unit/nva/search/RestHighLevelClientWrapper.java:74]: this class exists only to make the ElasticSearch client mockable. All methods in this class just call the respective EsClient method. In tests they are always mocked so they should be ignored by Jacoco

File: batch-index/src/main/java/no/unit/nva/search/BatchIndexer.java [hunk 6]
```diff
@@ -71,9 +72,11 @@ private ListingResult fetchNextPageOfFilenames() {
...
         logger.info("Indexing file:" + file.toString());
         Stream<JsonNode> fileContents = fetchFileContents(file);
-        Stream<Publication> publishedPublications = keepOnlyPublishedPublications(fileContents);
-        Stream<Try<SortableIdentifier>> indexActions = insertToIndex(publishedPublications);
-        List<String> failures = collectFailures(indexActions).collect(Collectors.toList());
+        List<IndexDocument> publishedPublications = keepOnlyPublishedPublications(fileContents)
+            .map(IndexDocument::fromPublication)
+            .collect(Collectors.toList());
+        List<BulkResponse> result = elasticSearchRestClient.batchInsert(publishedPublications);
+        List<String> failures = collectFailures(result).collect(Collectors.toList());
```
Comment (brinxmat) [inline batch-index/src/main/java/no/unit/nva/search/BatchIndexer.java:79]: I think it's a bit weird that the entire class simply returns a list of errors. For me, the IndexingResult class acts more like an interface for this class.

File: batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java [hunk 7]
```diff
@@ -14,21 +22,29 @@
     public static final String[] RESOURCES = {
         FIRST_FILE_HAS_NO_PUBLISHED_PUBLICATIONS,
         SECOND_FILE_HAS_PUBLISHED_PUBLICATIONS
-
     };
     public static final String EXPECTED_EXCEPTION_MESSAGE = "expectedMessage";
 
     public static final String FILE_WITH_IDENTIFIERS_OF_PUBLISHED_RESOURCES_IN_SAMPLE_FILES =
         "published_resources_identifiers_for_all_files.txt";
```
Comment (brinxmat) [inline batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java:29]: Remove unused vars

File: batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java [hunk 8]
```diff
@@ -14,21 +22,29 @@
...
     public static final Context CONTEXT = mock(Context.class);
     public static final String[] PUBLISHED_RESOURCES_IDENTIFIERS = readPublishedResourcesIdentifiersFromStaticFile();
+    public static final int ARBITRARY_LIMIT = 1000;
+    public static final Random RANDOM = new Random();
+    public static final int ARBITARY_QUERY_TIME = 123;
```
Comment (brinxmat) [inline batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java:34]: spelling

File: batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java [hunk 9]
```diff
@@ -37,4 +53,13 @@ public void addDocumentToIndex(IndexDocument document) throws SearchException {
         return IoUtils.linesfromResource(Path.of(FILE_WITH_IDENTIFIERS_OF_PUBLISHED_RESOURCES_IN_SAMPLE_FILES))
             .toArray(String[]::new);
     }
+
+    private Failure creteFailure(SortableIdentifier id) {
```
Comment (brinxmat) [inline batch-index/src/test/java/no/unit/nva/search/BatchIndexTest.java:57]: create

File: batch-index/src/test/java/no/unit/nva/search/StubElasticSearchHighLevelRestClient.java [hunk 10]
```diff
@@ -1,11 +1,19 @@
...
+import java.util.stream.Collectors;
 import no.unit.nva.search.exception.SearchException;
+import org.elasticsearch.action.DocWriteRequest.OpType;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkResponse;
 
 public class StubElasticSearchHighLevelRestClient extends ElasticSearchHighLevelRestClient {
 
+    public static final DocWriteResponse DOC_WRITE_RESPONSE = null;
```
Comment (brinxmat) [inline batch-index/src/test/java/no/unit/nva/search/StubElasticSearchHighLevelRestClient.java:15]: Remove unused

File: search-commons/src/test/java/no/unit/nva/search/ElasticsearchSigningHighLevelRestClientTest.java [hunk 11]
```diff
@@ -190,6 +197,32 @@ public void prepareIndexForBatchInsertCreatesIndexWhenResourcesIndexDoesNotExist
             .create(any(CreateIndexRequest.class), any(RequestOptions.class));
     }
 
+    @Test
+    public void batchInsertIndexesAllDocumentsInBatchInBulksOfSpecifiedSize() throws IOException {
+        RestHighLevelClientWrapper esClient = mock(RestHighLevelClientWrapper.class);
+        ElasticSearchHighLevelRestClient client = new ElasticSearchHighLevelRestClient(esClient);
+
+        var publications = IntStream.range(0, 1256).boxed()
```
Comment (brinxmat) [inline search-commons/src/test/java/no/unit/nva/search/ElasticsearchSigningHighLevelRestClientTest.java:205]: vars and typed mixing?

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**